src/schedule_synthesis.jl

```diff
@@ -0,0 +1,273 @@
+                break
+            end
+            meet += 1
+        end
+    end
+
+    safe_constraints
+end
+
+# Helper functions
+
+"""
+    _undigit(d[, base=2])
+
+Convert a list of digits to a number. Default base=2.
+```julia
+julia> _undigit([1, 0, 0])
+4
+```
+"""
+function _undigit(d::Vector{Int64}; base=2)::Int64
+    s = 0
+    mult = 1
+    for val in reverse(d)
+        s += val * mult
+        mult *= base
+    end
+    return s
+end
+
+"""
+    _digits_b2r(x[, pad])
+
+Digits **b**ase **2**, **r**everse
+A shortcut for `digits(x, base=2, pad=pad) |> reverse`
+"""
+function _digits_b2r(x::Int64, pad::Int64=0)::Vector{Int64}
+    digits(x, base=2, pad=pad) |> reverse
+end
+
+"""
+    _state_separation(l, B[, indigits=false])
+
+_state_separation takes a number `l` representing the overall state of
+a `_SynthesizedAutomaton` and an array `B` representing the number of bits
+for each comprising `Automaton` of the `_SynthesizedAutomaton` to compute a 
+list of individual states. For example
+
+```julia
+julia> _state_separation(6, [1, 2])
+[1, 2]
+```
+The explanation of the example is as follows
+```
+6 -> [1, 1, 0]
+[1, 1, 0] / [1, 2] -> [[1], [1, 0]]
+[[1], [1, 0]] -> [1, 2]
+```
+"""
+function _state_separation(l::Int64, B::Vector{Int64}; indigits=false)
+    @assert l < 2^sum(B) "l has more bits than the sum of B"
+    
+    bits = digits(l, base=2, pad=sum(B)) |> reverse
+    index_points = cumsum(B)
+    from_index = [0; index_points[1:end-1]]
+    to_index = index_points
+    bits_separated = [bits[i+1:j] for (i, j) in zip(from_index, to_index)]
+    if indigits
+        bits_separated
+    else
+        map(_undigit, bits_separated)
+    end
+end
+
+"""
+Struct for a controller automaton which enforces only the weakly 
+hard constraint for that controller. i.e., it is not concerned with the
+dynamical matrix of the system.
+"""
+struct _ConstraintAutomaton
+    # # of locations. Legal locations are in the range 0:L-1.
+    L::Int64
+    # # of actions. Legal actions are in the range 0:Σ-1.
+    Σ::Int64
+    # Transition function. T(l,σ) is a location in 0:L-1.
+    T::Function
+    # Initial location in L.
+    l_0::Int64
+    # Function that returns whether a location is final
+    Q_f::Function
+end
+
+"""
+Struct for a synthesized automaton from multiple constraint automata.
+"""
+struct _SynthesizedAutomaton
+    # # of comprising automata
+    N::Int64
+    # List of bits for all comprising controllers
+    B::Vector{Int64}
+    # Locations. Legal locations are in the range 0:L-1.
+    L::Int64
+    # List of actions.
+    Σ::Vector{Int64}
+    # Transition function.  T(l,σ) is a location in 0:L-1.
+    T::Function
+    # Initial location in L.
+    l_0::Int64
+    # Function that returns whether a location is final.
+    Q_f::Function
+end
+
+"""
+Build an constraint controller automaton for the given weakly hard constraint
+"""
+function _controller_automaton(c::WeaklyHardConstraint)
```

Comment:
This needs to specify the concrete type of constraint, since they have different members depending on type.

This could be called `_ConstraintAutomaton` to make it clearer what it's an external constructor for.